### PR TITLE
Refactor(plugins): Optimize loading of schemas

### PR DESCRIPTION
**Pull request by ClausHolbechArista**
## Change Summary

Optimize loading of schemas in plugins

## Component(s) name

`arista.avd.yaml_templates_to_facts`
`arista.avd.eos_designs_facts`
`arista.avd.validate_and_template`
`arista.avd.convert_schema`

## Proposed changes

- Add id to builtin schemas
- Preloading built-in schemas in a "store"
- Updating schema tools and plugins to load schema using id instead of loading the full schema in ansible and passing it as argument.
- Update `convert_schema` to also load schema with id.
- Various optimizations
- Fix various issues when using $ref
- Test $ref between separate schemas

## How to test

No changes to molecule.
Tested with $ref to other schema - see Test.md in first commit.

### CODE REVIEW

**Review by carlbuchmann — APPROVED** [review]
LGTM tested with PR #2517

File: ansible_collections/arista/avd/plugins/action/eos_designs_facts.py [hunk 1]
```diff
@@ -28,9 +28,10 @@ def run(self, tmp=None, task_vars=None):
             profiler = cProfile.Profile()
             profiler.enable()
 
+        self._plugin_name = task_vars["ansible_role_name"]
```
Comment (gmuloc) [inline ansible_collections/arista/avd/plugins/action/eos_designs_facts.py:31]: any reason to have plugin_name different from schema_id?

File: ansible_collections/arista/avd/plugins/action/validate_and_template.py [hunk 2]
```diff
@@ -19,16 +19,12 @@ def run(self, tmp=None, task_vars=None):
...
-            if not isinstance(schema, dict):
-                raise AnsibleActionFail("The argument 'schema' must be a dict")
-        else:
-            raise AnsibleActionFail("The arguments 'template' and 'schema' must be set")
+        self.templatefile = self._task.args["template"]
+        if not isinstance(self.templatefile, str):
+            raise AnsibleActionFail("The argument 'template' must be a string")
 
+        schema = self._task.args.get("schema")
+        schema_id = self._task.args.get("schema_id")
```
Comment (gmuloc) [inline ansible_collections/arista/avd/plugins/action/validate_and_template.py:27]: so here we lose the AnsibleActionFail to state that the argument must be set and a dict?

File: ansible_collections/arista/avd/plugins/action/validate_and_template.py [hunk 3]
```diff
@@ -19,16 +19,12 @@ def run(self, tmp=None, task_vars=None):
...
-        if self._task.args and "schema" in self._task.args and "template" in self._task.args:
-            self.templatefile = self._task.args["template"]
-            if not isinstance(self.templatefile, str):
-                raise AnsibleActionFail("The argument 'template' must be a string")
-            schema = self._task.args["schema"]
-            if not isinstance(schema, dict):
-                raise AnsibleActionFail("The argument 'schema' must be a dict")
-        else:
-            raise AnsibleActionFail("The arguments 'template' and 'schema' must be set")
+        self.templatefile = self._task.args["template"]
```
Comment (gmuloc) [inline ansible_collections/arista/avd/plugins/action/validate_and_template.py]: no more gracefully raising if template is not present?

File: ansible_collections/arista/avd/plugins/action/yaml_templates_to_facts.py [hunk 4]
```diff
@@ -55,13 +55,15 @@ def run(self, tmp=None, task_vars=None):
...
             template_output = self._task.args.get("template_output", False)
             debug = self._task.args.get("debug", False)
             remove_avd_switch_facts = self._task.args.get("remove_avd_switch_facts", False)
             conversion_mode = self._task.args.get("conversion_mode")
             validation_mode = self._task.args.get("validation_mode")
             output_schema = self._task.args.get("output_schema")
+            output_schema_id = self._task.args.get("output_schema_id")
 
         else:
             raise AnsibleActionFail("The argument 'templates' must be set")
```
Comment (gmuloc) [inline ansible_collections/arista/avd/plugins/action/yaml_templates_to_facts.py:69]: this looks like a weird error message for all that could go wrong:)

File: ansible_collections/arista/avd/plugins/action/yaml_templates_to_facts.py [hunk 5]
```diff
@@ -55,13 +55,15 @@ def run(self, tmp=None, task_vars=None):
                 raise AnsibleActionFail("The argument 'templates' must be set as a list")
 
             schema = self._task.args.get("schema")
+            schema_id = self._task.args.get("schema_id")
```
Comment (gmuloc) [inline ansible_collections/arista/avd/plugins/action/yaml_templates_to_facts.py:58]: should these be documented in module?
https://github.com/aristanetworks/ansible-avd/blob/devel/ansible_collections/arista/avd/plugins/modules/yaml_templates_to_facts.py#L29

File: ansible_collections/arista/avd/plugins/action/yaml_templates_to_facts.py [hunk 6]
```diff
@@ -177,7 +195,7 @@ def run(self, tmp=None, task_vars=None):
 
                 # If output_schema is set, perform inplace conversion of each returned dict according to output_schema
                 # to normalize the data to correct format before merging
-                if output_schema:
+                if output_avdschema:
```
Comment (gmuloc) [inline ansible_collections/arista/avd/plugins/action/yaml_templates_to_facts.py:198]: should we change the comment above? - I am not sure if this change fixes some unworking thing?

File: ansible_collections/arista/avd/plugins/filter/convert_schema.py [hunk 7]
```diff
@@ -29,16 +29,16 @@ def convert_schema(schema: dict, type: str):
     AvdSchemaError, AvdValidationError
         If the input schema is not valid, exceptions will be raised accordingly.
     """
-    avdschema = AvdSchema(schema)
+    avdschema = AvdSchema(schema_id=schema_id)
```
Comment (gmuloc) [inline ansible_collections/arista/avd/plugins/filter/convert_schema.py:34]: Should we change the `Returns` section in the doctsring?

File: ansible_collections/arista/avd/plugins/plugin_utils/schema/avdschema.py [hunk 8]
```diff
@@ -26,15 +27,23 @@
...
         if JSONSCHEMA_IMPORT_ERROR:
             raise AristaAvdError('Python library "jsonschema" must be installed to use this plugin') from JSONSCHEMA_IMPORT_ERROR
         if DEEPMERGE_IMPORT_ERROR:
             raise AristaAvdError('Python library "deepmerge" must be installed to use this plugin') from DEEPMERGE_IMPORT_ERROR
 
+        self.store = create_store()
         if not schema:
-            schema = DEFAULT_SCHEMA
-        self._schema_validator = jsonschema.Draft7Validator(AVD_META_SCHEMA)
+            if schema_id:
```
Comment (gmuloc) [inline ansible_collections/arista/avd/plugins/plugin_utils/schema/avdschema.py]: probably worth adding a comment here to explain the order of precedence
supplied schema > schema_id > DEFAULT_SCHEMA

File: ansible_collections/arista/avd/plugins/plugin_utils/schema/avdschemaresolver.py [hunk 9]
```diff
@@ -1,10 +1,12 @@
 from __future__ import absolute_import, annotations, division, print_function
 
+from ansible_collections.arista.avd.plugins.plugin_utils.schema.refresolver import create_refresolver
```
Comment (gmuloc) [inline ansible_collections/arista/avd/plugins/plugin_utils/schema/avdschemaresolver.py]: I would expect isort to be unhappy about this but it seems not?

File: ansible_collections/arista/avd/plugins/plugin_utils/schema/avdtodocumentationschemaconverter.py [hunk 10]
```diff
@@ -105,7 +105,7 @@ def convert_schema(self):
         filenames = self._get_filenames(schema)
 
         for filename in filenames:
-            output.append({"filename": filename, "tables": self.build_tables(filename, schema)})
+            output[filename] = {"filename": filename, "tables": self.build_tables(filename, schema)}
```
Comment (gmuloc) [inline ansible_collections/arista/avd/plugins/plugin_utils/schema/avdtodocumentationschemaconverter.py]: This looks a bit weird to have filename both as the key and inside the dictionary - should filename be removed (cf documentation where it is)

File: ansible_collections/arista/avd/plugins/action/eos_designs_facts.py (hunk 1 above)
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/plugins/action/eos_designs_facts.py:31]: We don't know which schema name is used when calling this. So I don't think we should hardcode a relation there.

File: ansible_collections/arista/avd/plugins/action/validate_and_template.py (hunk 2 above)
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/plugins/action/validate_and_template.py:27]: It is raised in the schematools like someone (!) told me to do when we built conversion_mode etc.

File: ansible_collections/arista/avd/plugins/action/validate_and_template.py (hunk 3 above)
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/plugins/action/validate_and_template.py]: changed to get()

File: ansible_collections/arista/avd/plugins/action/yaml_templates_to_facts.py (hunk 4 above)
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/plugins/action/yaml_templates_to_facts.py:69]: it's only for the case of no arguments.

File: ansible_collections/arista/avd/plugins/action/yaml_templates_to_facts.py (hunk 6 above)
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/plugins/action/yaml_templates_to_facts.py:198]: I just check on a var that is set if either output_schema or output_schema_it is set.

File: ansible_collections/arista/avd/plugins/filter/convert_schema.py (hunk 7 above)
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/plugins/filter/convert_schema.py:34]: updated to `dict | list`

File: ansible_collections/arista/avd/plugins/plugin_utils/schema/avdtodocumentationschemaconverter.py (hunk 10 above)
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/plugins/plugin_utils/schema/avdtodocumentationschemaconverter.py]: it needs to be both places

File: ansible_collections/arista/avd/plugins/action/yaml_templates_to_facts.py (hunk 5 above)
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/plugins/action/yaml_templates_to_facts.py:58]: updated docs

File: ansible_collections/arista/avd/plugins/plugin_utils/schema/avdschema.py (hunk 8 above)
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/plugins/plugin_utils/schema/avdschema.py]: added docstring

File: ansible_collections/arista/avd/plugins/filter/convert_schema.py [hunk 11]
```diff
@@ -8,37 +8,37 @@
...
-        Input AVD Schema
+    schema_id : str, ["eos_cli_config_gen" , "eos_designs"]
+        ID of AVD Schema
     type : str, ["documentation", "jsonschema"]
         Type of schema to convert to
 
     Returns
     -------
-    dict
+    dict | list
```
Comment (gmuloc) [inline ansible_collections/arista/avd/plugins/filter/convert_schema.py:26]: I am wondering if this is very good to have a return type being either a list or a dict now.. should we maybe separate in two separate functions for more clarity?

File: ansible_collections/arista/avd/plugins/filter/convert_schema.py [hunk 12]
```diff
@@ -8,37 +8,37 @@
...
-    avdschema = AvdSchema(schema)
+    avdschema = AvdSchema(schema_id=schema_id)
     if type == "documentation":
-        schemaconverter = AvdToDocumentationSchemaConverter(avdschema)
+        return AvdToDocumentationSchemaConverter(avdschema).convert_schema()
+
     elif type == "jsonschema":
-        schemaconverter = AvdToJsonSchemaConverter(avdschema)
+        return AvdToJsonSchemaConverter(avdschema).convert_schema()
+
```
Comment (gmuloc) [inline ansible_collections/arista/avd/plugins/filter/convert_schema.py:40]: just noticed that the error message below is not really correct (not mentioning `jsonschema` type)

File: ansible_collections/arista/avd/plugins/plugin_utils/schema/avdschema.py [hunk 13]
```diff
@@ -26,15 +27,36 @@
...
-    def __init__(self, schema: dict = None):
+    """
+    AvdSchema takes either a schema as dict or the ID of a builtin schema.
+    If none of them are set, a default "dummy" schema will be loaded.
+    schema -> schema_id -> DEFAULT_SCHEMA
+
+    Parameters
+    ----------
+    schema : dict
+        AVD Schema as dictionary. Will be verified towards avd_meta_schema.
```
Comment (gmuloc) [inline ansible_collections/arista/avd/plugins/plugin_utils/schema/avdschema.py]:
```suggestion
        AVD Schema as dictionary. Will be verified towards AVD_META_SCHEMA.
```

**Review by gmuloc — DISMISSED** [review]
couple of new comments - otherwise it looks good to me

File: ansible_collections/arista/avd/plugins/filter/convert_schema.py (hunk 12 above)
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/plugins/filter/convert_schema.py:40]: fixed

File: ansible_collections/arista/avd/plugins/filter/convert_schema.py (hunk 11 above)
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/plugins/filter/convert_schema.py:26]: I suspect there will be many changes to the documentation area as we get through the eos_designs schemas, so I will split it next time we have to rework it. Adding TODO for now.

File: ansible_collections/arista/avd/plugins/plugin_utils/schema/avdschema.py (hunk 13 above)
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/plugins/plugin_utils/schema/avdschema.py]: fixed

**Review by gmuloc — APPROVED** [review]

**Review by carlbuchmann — APPROVED** [review]
LGTM - tested with PR #2517

**Review by gmuloc — APPROVED** [review]

**Pull request merged**